src/Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs

```diff
@@ -0,0 +1,408 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Globalization;
+using System.Text;
+
+namespace Microsoft.Net.Http.Headers
+{
+    public class MediaTypeHeaderValue
+    {
+        private const string CharsetString = "charset";
+        private const string BoundaryString = "boundary";
+
+        private static readonly HttpHeaderParser<MediaTypeHeaderValue> SingleValueParser
+            = new GenericHeaderParser<MediaTypeHeaderValue>(false, GetMediaTypeLength);
+        private static readonly HttpHeaderParser<MediaTypeHeaderValue> MultipleValueParser
+            = new GenericHeaderParser<MediaTypeHeaderValue>(true, GetMediaTypeLength);
+
+        // Use list instead of dictionary since we may have multiple parameters with the same name.
+        private ICollection<NameValueHeaderValue> _parameters;
+        private string _mediaType;
+
+        private MediaTypeHeaderValue()
+        {
+            // Used by the parser to create a new instance of this type.
+        }
+
+        public MediaTypeHeaderValue(string mediaType)
+        {
+            CheckMediaTypeFormat(mediaType, "mediaType");
+            _mediaType = mediaType;
+        }
+
+        public MediaTypeHeaderValue(string mediaType, double quality)
+            : this(mediaType)
+        {
+            Quality = quality;
+        }
+
+        public string Charset
+        {
+            get
+            {
+                return NameValueHeaderValue.Find(_parameters, CharsetString)?.Value;
+            }
+            set
+            {
+                // We don't prevent a user from setting whitespace-only charsets. Like we can't prevent a user from
+                // setting a non-existing charset.
+                var charsetParameter = NameValueHeaderValue.Find(_parameters, CharsetString);
+                if (string.IsNullOrEmpty(value))
+                {
+                    // Remove charset parameter
+                    if (charsetParameter != null)
+                    {
+                        _parameters.Remove(charsetParameter);
+                    }
+                }
+                else
+                {
+                    if (charsetParameter != null)
+                    {
+                        charsetParameter.Value = value;
+                    }
+                    else
+                    {
+                        Parameters.Add(new NameValueHeaderValue(CharsetString, value));
+                    }
+                }
+            }
+        }
+
+        public Encoding Encoding
```

Comment:
Note to callers: Don't trust the Encoding coming from the client.  Enforce UTF8 or similar.  Do we need enforcement at this API level or in the caller (MVC)?